File: 547-number-of-provinces/number-of-provinces.cpp
class Solution {
public:
    void dfs(unordered_map<int, vector<int>>& adj, int node, vector<bool> &visited){
        for(int i : adj[node]){
            if(!visited[i]){
                visited[i] = true;
                dfs(adj, i, visited);
            }
        }
    }
    int findCircleNum(vector<vector<int>>& isConnected) {
        int n = isConnected.size();
        vector<bool> visited(n,false);
        int ans = 0;
        unordered_map<int, vector<int> > adj;
        for(int i=0; i<isConnected.size(); i++){
            for(int j=0; j<isConnected[i].size(); j++){
                if(isConnected[i][j] == 1){
                        adj[i].push_back(j);
                        adj[j].push_back(i);
                }
            }
        }
        for(int i=0; i<n; i++){
            if(!visited[i]){
                ans++;
                dfs(adj, i, visited);
                visited[i] = true;

            }
        }
        return ans;
    }
};

File: 547-number-of-provinces/number-of-provinces.java
class Solution {
    public int findCircleNum(int[][] isConnected) {
        int totalNumberOfProvinces = 0;
        int n = isConnected.length;

        Deque<Integer> dq = new ArrayDeque<>();

        boolean[] visited = new boolean[n];

        for(int i = 0; i < n; i++){
            if(!visited[i]){
                totalNumberOfProvinces++;
                dq.add(i);
                while(!dq.isEmpty()){
                    int node = dq.pollFirst();
                    visited[node] = true;

                    for(int j = 0; j < n; j++){
                        if(isConnected[node][j] == 1 && !visited[j])
                            dq.add(j);
                    }
                }
            }
        }

        return totalNumberOfProvinces;
    }
}